### PR TITLE
Introduce cheriot subarch and cheriotrtos in target triples.

### PR DESCRIPTION
**Pull request by resistor**
- **[CHERIoT] Introduce a cheriot subarch of RV32**
- **[CHERIoT] Add cheriotrtos as an OS in the triple**

### CODE REVIEW

**Comment by resistor** [comment]
@davidchisnall Would like confirmation that this is good with you before it goes in.

**Comment by davidchisnall** [comment]
It might be good to put a v1 in there, so we can follow the same structure for v2.

File: clang/lib/Driver/ToolChains/Arch/RISCV.cpp [hunk 1]
```diff
@@ -239,6 +240,13 @@ StringRef riscv::getRISCVABI(const ArgList &Args, const llvm::Triple &Triple) {
   // rv64g | rv64*d -> lp64d
   // rv64* -> lp64
   StringRef Arch = getRISCVArch(Args, Triple);
+  if (Triple.getSubArch() == llvm::Triple::RISCV32SubArch_cheriot) {
+    llvm::Triple::OSType OS = Triple.getOS();
+    if (OS == llvm::Triple::CheriotRTOS)
+      return "cheriot";
+    else if (OS == llvm::Triple::UnknownOS)
+      return "cheriot-baremetal";
```
Comment (davidchisnall) [inline clang/lib/Driver/ToolChains/Arch/RISCV.cpp:248]: What ABI are we getting for all of the ‘unknown’ things in the tests?

**Review by davidchisnall — COMMENTED** [review]
This will also want to bump the CHERIoT macro define so that we can make sure that people have a sufficiently new compiler for the build system to work.

File: clang/lib/Driver/ToolChains/Arch/RISCV.cpp (hunk 1 above)
Comment (resistor) [reply clang/lib/Driver/ToolChains/Arch/RISCV.cpp:248]: Not totally sure I follow the question, but riscv32cheriot-unknown-unknown will imply "cheriot-baremetal" ABI. I think this best matches user expectations in that "i386-unknown-unknown" would be the target triple for bare metal i386, etc.

**Comment by resistor** [comment]
> It might be good to put a v1 in there, so we can follow the same structure for v2.

I made the enum include v1. The CLI parser will accept "riscv32cheriot" or "riscv32cheriotv1".

> This will also want to bump the CHERIoT macro define so that we can make sure that people have a sufficiently new compiler for the build system to work.

Done

**Comment by davidchisnall** [comment]
LGTM.  I think we'll need some extra xmake good to ensure that the build system provides the correct arguments for the old and new compiler.  Not sure the best way of doing that currently.

**Pull request merged**